### PR TITLE
GEODE-9947: Add LINDEX command

### PR DESCRIPTION
**Pull request by Kris-10-0**
### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [x] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [x] Is your initial contribution a single, squashed commit?

- [x] Does `gradlew build` run cleanly?

- [x] Have you written or updated unit tests to verify your changes?

- [NA] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### CODE REVIEW

File: geode-docs/tools_modules/geode_for_redis.html.md.erb [hunk 1]
```diff
@@ -186,7 +186,7 @@ Could not connect to Redis at 127.0.0.1:6379: Connection refused
 | HINCRBY | HINCRBYFLOAT | HKEYS | HLEN |
 | HMGET | HMSET | HSCAN **[3]** | HSET |
 | HSETNX | HSTRLEN | HVALS | INCR |
-| INCRBY | INCRBYFLOAT | INFO **[4]** | KEYS |
+| INCRBY | INCRBYFLOAT | INFO **[4]** | KEYS | LINDEX |
```
Comment (DonalEvans) [inline geode-docs/tools_modules/geode_for_redis.html.md.erb:189]: I think that the formatting of this section will get messed up if you just add another element like this, since it's currently set up as a table with 4 columns. Adding LINDEX here will (I think) add a fifth column with only LINDEX in it. This makes updating the list of commands a nightmare, since adding one means you need to adjust every command that comes after it in the list. We probably need a better way of handling this section if we're planning to modify it frequently. Maybe breaking it up by command type? That's work for another ticket though, I think.

Luckily (for this PR, at least) if you add LPUSH, LPOP and LLEN to the list of supported commands too, then not only will the docs be up to date, but you can just add one extra row of 4 elements below this one and not have to worry about affecting the rest of the table.

File: geode-for-redis/src/main/java/org/apache/geode/redis/internal/data/RedisList.java [hunk 2]
```diff
@@ -42,6 +42,24 @@ public RedisList() {
     this.elementList = new SizeableByteArrayList();
   }
 
+  /**
+   * @param index index of desired element. Positive index starts at the head. Negative index starts
+   *        at the tail.
+   * @return element at index. nil if index is out of range.
```
Comment (DonalEvans) [inline geode-for-redis/src/main/java/org/apache/geode/redis/internal/data/RedisList.java:48]: This should be "null" rather than "nil".

File: geode-for-redis/src/integrationTest/java/org/apache/geode/redis/internal/commands/executor/list/AbstractLIndexIntegrationTest.java [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+  public void lindex_withPositiveIndex_returnsElemnt() {
+    jedis.lpush(LIST_KEY, LIST_ELEMENTS);
+    assertThat(jedis.lindex(LIST_KEY, INDEX)).isEqualTo("deer");
+  }
+
+  @Test
+  public void lindex_withNegativeIndex_returnsElemnt() {
+    jedis.lpush(LIST_KEY, LIST_ELEMENTS);
+    assertThat(jedis.lindex(LIST_KEY, -INDEX)).isEqualTo("chameleon");
+  }
```
Comment (DonalEvans) [inline geode-for-redis/src/integrationTest/java/org/apache/geode/redis/internal/commands/executor/list/AbstractLIndexIntegrationTest.java:81]: Could these tests be expanded to check every element in the list, just to cover the weird corner case where somehow we always return the same value regardless of index?

File: geode-for-redis/src/integrationTest/java/org/apache/geode/redis/internal/commands/executor/list/AbstractLIndexIntegrationTest.java [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+  public void ensureListConsistency_whenRunningConcurrently() {
+    final AtomicReference<String> lindexResultReference = new AtomicReference<>();
+    new ConcurrentLoopingThreads(1000,
+        i -> jedis.lpush(LIST_KEY, LIST_ELEMENTS),
+        i -> lindexResultReference.set(jedis.lindex(LIST_KEY, -7)))
+            .runWithAction(() -> {
+              assertThat(lindexResultReference).satisfiesAnyOf(
+                  lindexResult -> assertThat(lindexResult.get()).isNull(),
+                  lindexResult -> assertThat(lindexResult.get()).isEqualTo("goat"));
+              jedis.del(LIST_KEY);
```
Comment (DonalEvans) [inline geode-for-redis/src/integrationTest/java/org/apache/geode/redis/internal/commands/executor/list/AbstractLIndexIntegrationTest.java:133]: This test might be a little bit better if instead of checking the concurrency behaviour between creating a new list via LPUSH and checking the index via LINDEX, it was checking the behaviour when adding a different set of elements to an already existing list, and checking that the last element is as expected. By checking the last element, we ensure that the index is never out of range, so we're more likely to catch any issues where we're reading the list while still adding elements to it.

**Review by DonalEvans — CHANGES_REQUESTED** [review]
Looks really good, just a few small suggestions.

File: geode-for-redis/src/integrationTest/java/org/apache/geode/redis/internal/commands/executor/list/AbstractLIndexIntegrationTest.java (hunk 4 above)
Comment (Kris-10-0) [reply geode-for-redis/src/integrationTest/java/org/apache/geode/redis/internal/commands/executor/list/AbstractLIndexIntegrationTest.java]: With the commands that are currently implemented, the only remove is lpop, which removes one element from the head. So I added only one element. I could change it so I add more then just a for loop to pop multiple added elements. 

When you add to the list it gets added to the head, so I checked the 0 index for the difference. If I checked the last element then for one of the cases it could be out of range.

File: geode-for-redis/src/integrationTest/java/org/apache/geode/redis/internal/commands/executor/list/AbstractLIndexIntegrationTest.java [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+    assertThat(jedis.lindex(NON_EXISTENT_LIST_KEY, -2)).isNull();
+  }
+
+  @Test
+  public void lindex_withPositiveIndexes_returnsElement() {
+    jedis.lpush(LIST_KEY, LIST_ELEMENTS);
+    for (int i = 0; i < LIST_ELEMENTS.length; i++) {
+      // LIST_ELEMENTS[LIST_ELEMENTS.length - 1 - i] iterates LIST_ELEMENTS backwards
+      System.out
+          .println(jedis.lindex(LIST_KEY, i) + " " + LIST_ELEMENTS[LIST_ELEMENTS.length - 1 - i]);
```
Comment (DonalEvans) [inline geode-for-redis/src/integrationTest/java/org/apache/geode/redis/internal/commands/executor/list/AbstractLIndexIntegrationTest.java:76]: This should be removed.

File: geode-for-redis/src/integrationTest/java/org/apache/geode/redis/internal/commands/executor/list/AbstractLIndexIntegrationTest.java [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+      assertThat(jedis.lindex(LIST_KEY, i)).isEqualTo(LIST_ELEMENTS[LIST_ELEMENTS.length - 1 - i]);
+    }
+  }
+
+  @Test
+  public void lindex_withNegativeIndexes_returnsElement() {
+    jedis.lpush(LIST_KEY, LIST_ELEMENTS);
+    for (int i = -7; i < 0; i++) {
+      // LIST_ELEMENTS[-(i + 1)] iterates LIST_ELEMENTS forwards
+      assertThat(jedis.lindex(LIST_KEY, i)).isEqualTo(LIST_ELEMENTS[-(i + 1)]);
```
Comment (DonalEvans) [inline geode-for-redis/src/integrationTest/java/org/apache/geode/redis/internal/commands/executor/list/AbstractLIndexIntegrationTest.java:86]: For consistency with the above test, and to make it a little easier to parse, could this instead be:
```
    for (int i = 0; i < LIST_ELEMENTS.length; i++) {
      assertThat(jedis.lindex(LIST_KEY, -(i + 1))).isEqualTo(LIST_ELEMENTS[i]);
    }
```

File: geode-for-redis/src/integrationTest/java/org/apache/geode/redis/internal/commands/executor/list/AbstractLIndexIntegrationTest.java [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+    String elementToAdd = "zebra";
+    final AtomicReference<String> lindexResultReference = new AtomicReference<>();
+    new ConcurrentLoopingThreads(1000,
+        i -> jedis.lpush(LIST_KEY, elementToAdd),
+        i -> lindexResultReference.set(jedis.lindex(LIST_KEY, 0)))
+            .runWithAction(() -> {
+              assertThat(lindexResultReference).satisfiesAnyOf(
+                  lindexResult -> assertThat(lindexResult.get()).isEqualTo(elementToAdd),
+                  lindexResult -> assertThat(lindexResult.get()).isEqualTo("goat"));
+              jedis.lpop(LIST_KEY);
```
Comment (DonalEvans) [inline geode-for-redis/src/integrationTest/java/org/apache/geode/redis/internal/commands/executor/list/AbstractLIndexIntegrationTest.java:143]: Good call on checking index 0 rather than the last index, my brain was thinking about things backwards.

However, it's not possible to spot concurrency issues when only pushing one element (the element is either there or it's not, so we can't ever tell if we tried to get the element at index 0 while elements were still being added), so for this test, a better approach would be to add multiple new elements, check index 0, then delete the whole key and re-add the initial elements at the end of the `runWithAction()` to get us back to the starting conditions.

**Review by DonalEvans — CHANGES_REQUESTED** [review]

**Review by DonalEvans — APPROVED** [review]

**Review by onichols-pivotal — COMMENTED** [review]
As per https://cwiki.apache.org/confluence/display/GEODE/Commit+Message+Format, first commit summary should be worded imperatively and  additional detail paragraph(s) should describe the problem and how this change solves it.

By amending your initial commit and force-pushing now, you can benefit from review feedback on your commit message, and you won't have to remember to fix it later when you merge.

**Review by onichols-pivotal — COMMENTED** [review]
As per https://cwiki.apache.org/confluence/display/GEODE/Commit+Message+Format, first commit summary should be meaningful and  additional detail paragraph(s) should describe the problem and how this change solves it.

By amending your initial commit and force-pushing now, you can benefit from review feedback on your commit message, and you won't have to remember to fix it later when you merge.

**Review by onichols-pivotal — COMMENTED** [review]
As per https://cwiki.apache.org/confluence/display/GEODE/Commit+Message+Format, first commit summary should be meaningful and  details should be linewrapped to a max width of 72 characters (your longest detail line is 255 characters) (use https://www.browserling.com/tools/word-wrap or `git config --global core.editor "vim \"+set textwidth=72\""` or enable IntelliJ Commit message inspections).

By amending your initial commit and force-pushing now, you can benefit from review feedback on your commit message, and you won't have to remember to fix it later when you merge.

File: geode-for-redis/src/main/java/org/apache/geode/redis/internal/commands/executor/list/LIndexExecutor.java [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+  public RedisResponse executeCommand(Command command, ExecutionHandlerContext context) {
+    List<byte[]> commandElems = command.getProcessedCommand();
+    RedisKey key = command.getKey();
+
+    return context.listLockedExecute(key, true,
+        list -> preformCommand(commandElems.get(2), list));
+  }
+
+  // Used to not throw an error if the list does not exist and has an invalid index
+  private RedisResponse preformCommand(byte[] unprocessedIndex, RedisList list) {
```
Comment (jdeppe-pivotal) [inline geode-for-redis/src/main/java/org/apache/geode/redis/internal/commands/executor/list/LIndexExecutor.java:41]: Minor typo on the method name, but not critical.

**Review by jdeppe-pivotal — APPROVED** [review]
LGTM

**Review by onichols-pivotal — COMMENTED** [review]
As per https://cwiki.apache.org/confluence/display/GEODE/Commit+Message+Format, first commit summary should be meaningful and details should be linewrapped to a max width of 72 characters (your longest detail line is 255 characters) (use https://www.browserling.com/tools/word-wrap or `git config --global core.editor "vim \"+set textwidth=72\""` or enable IntelliJ Commit message inspections).

By amending your initial commit and force-pushing now, you can benefit from review feedback on your commit message, and you won't have to remember to fix it later when you merge.

**Review by onichols-pivotal — COMMENTED** [review]
Thanks for improving your commit message!  Don't forget, first commit details should be linewrapped to a max width of 72 characters (your longest detail line is 255 characters) (use https://www.browserling.com/tools/word-wrap or `git config --global core.editor "vim \"+set textwidth=72\""` or enable IntelliJ Commit message inspections).
